src/cli/self_update/shell.rs

```diff
@@ -0,0 +1,192 @@
+    // Used primarily in checking rcfiles for cleanup.
+    fn rcfiles(&self) -> Vec<PathBuf>;
+
+    // Gives rcs that should be written to.
+    fn update_rcs(&self) -> Vec<PathBuf>;
+
+    // Writes the relevant env file.
+    fn env_script(&self) -> ShellScript {
+        ShellScript {
+            name: "env",
+            content: POSIX_ENV,
+        }
+    }
+
+    fn source_string(&self) -> Result<String> {
+        Ok(format!(r#"source "{}/env""#, canonical_cargo_home()?))
+    }
+}
+
+struct Posix;
+impl UnixShell for Posix {
+    fn does_exist(&self) -> bool {
+        true
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        match utils::home_dir() {
+            Some(dir) => vec![dir.join(".profile")],
+            _ => vec![],
+        }
+    }
+
+    fn update_rcs(&self) -> Vec<PathBuf> {
+        // Write to .profile even if it doesn't exist.
+        self.rcfiles()
+    }
+}
+
+struct Bash;
+
+// Bash is the source of many complications because it loads ONLY 1 rc,
+// either a login profile or .bashrc, not both.
+impl Bash {
+    // Bash will load only one of these, the first one that exists, when
+    // a login shell starts. BUT not all shells start inside a login shell!
+    fn profiles() -> Vec<PathBuf> {
+        [".bash_profile", ".bash_login", ".profile"]
+            .iter()
+            .filter_map(|rc| utils::home_dir().map(|dir| dir.join(rc)))
+            .collect()
+    }
+
+    // Bash will load only .bashrc on the start of most GUI terminals.
+    fn rc() -> Option<PathBuf> {
+        utils::home_dir().map(|dir| dir.join(".bashrc"))
+    }
+}
+
+impl UnixShell for Bash {
+    fn does_exist(&self) -> bool {
+        matches!(process().var("SHELL"), Ok(sh) if sh.contains("bash"))
+            || self.rcfiles().iter().any(|rc| rc.is_file())
+            || matches!(utils::find_cmd(&["bash"]), Some(_))
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        let mut profiles = Bash::profiles();
+        if let Some(rc) = Bash::rc() {
+            profiles.push(rc);
+        }
+        profiles
+    }
+
+    fn update_rcs(&self) -> Vec<PathBuf> {
+        Bash::profiles()
+            .into_iter()
+            .filter(|rc| rc.is_file())
+            // bash only reads one "login profile" so pick the one that exists
+            .take(1)
+            // Pick .bashrc if it exists for GUI terminals
+            .chain(Bash::rc().filter(|rc| rc.is_file()))
+            .collect()
+    }
+}
+
+struct Zsh;
+impl UnixShell for Zsh {
+    fn does_exist(&self) -> bool {
+        matches!(process().var("SHELL"), Ok(sh) if sh.contains("zsh"))
+            || matches!(process().var("ZDOTDIR"), Ok(dir) if dir.len() > 0)
+            || self.rcfiles().iter().any(|rc| rc.is_file())
+            || matches!(utils::find_cmd(&["zsh"]), Some(_))
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        // FIXME: if zsh exists but is not in the process tree of the shell
```

Comment:
This seems like an intrinsic characteristic unless we get into very convoluted territory where we invoke rustup inside zsh to read out the calculated value of ZDOTDIR.

I think we should either plan on doing that in a future iteration, or not have a FIXME and instead have a FAQ about limitations.

Reply:
Shell out to zsh and ask it to report back? :laughing:

Reply:
yeah - assuming zsh accepts -c , something like `zsh -c "{ourargv0} report-env ZDOTDIR"`.

Reply:
It does, currently thinking `zsh -c "echo $ZDOTDIR"` or a close variation would be simplest from what I remember of all the APIs involved.

Reply:
I have a draft of this logic implemented and I've done some basic tinkering around to verify that it should work. Writing tests for it is going to be... interesting.